File: Arrays & Hashing/287. Find the Duplicate Number/Solution.java
import java.lang.Math;

class Solution {
    public int findDuplicate(int[] nums) {
        // N : len(nums)
        // Time Complexity: O(N logN)
        // Space Complexity: O(1)
        // int n = nums.length;
        // Arrays.sort(nums);
        // for (int i = 1; i < n; i++) {
        //     if (nums[i] == nums[i - 1]) return nums[i];
        // }
        // return -1;


        // N : len(nums)
        // Time Complexity: O(N)
        // Space Complexity: O(N)
        // int n = nums.length;
        // int[] freq = new int[n];
        // for (int num: nums) {
        //     if (freq[num] != 0) return num;
        //     freq[num]++;
        // }
        // return -1;


        // N : len(nums)
        // Time Complexity: O(N)
        // Space Complexity: O(1)
        int n = nums.length;
        for (int i = 0; i < n; i++) {
            if (nums[Math.abs(nums[i]) - 1] < 0) return Math.abs(nums[i]);
            nums[Math.abs(nums[i]) - 1] *= -1;
        }
        return -1;
    }
}

File: Arrays & Hashing/287. Find the Duplicate Number/Solution.py
class Solution:
    def findDuplicate(self, nums: list[int]) -> int:
        # N : len(nums)
        # Time Complexity: O(N logN)
        # Space Complexity: O(1)
        # n = len(nums)
        # nums.sort()
        # for i in range(1, n):
        #     if nums[i] == nums[i - 1]:
        #         return nums[i]
        # return -1


        # N : len(nums)
        # Time Complexity: O(N)
        # Space Complexity: O(N)
        # n = len(nums)
        # freq = [0] * n
        # for num in nums:
        #     if freq[num] == 1:
        #         return num
        #     freq[num] += 1
        # return -1

        # N : len(nums)
        # Time Complexity: O(N)
        # Space Complexity: O(1)
        n = len(nums)
        for i in range(n):
            if nums[abs(nums[i]) - 1] < 0:
                return abs(nums[i])
            nums[abs(nums[i]) - 1] *= -1
        return -1